openid/openid.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2016 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENSE file for details.
+
+// Package openid contains functions to help log-in to Ubuntu SSO using
+// OpenID 2.0.
+package openid
+
+import (
+	"net/url"
+	"strings"
+
+	"github.com/yohcop/openid-go"
+	"gopkg.in/errgo.v1"
+
+	"github.com/juju/usso"
+)
+
+const (
+	// These standard simple registration fields are supported by
+	// Ubuntu SSO.
+	SRegNickname = "nickname"
+	SRegEmail    = "email"
+	SRegFullName = "fullname"
+	SRegPostcode = "postcode"
+	SRegCountry  = "country"
+	SRegLanguage = "language"
+	SRegTimezone = "timezone"
+
+	// These non-standard simple registration fields are supported by
+	// Ubuntu SSO.
+	SRegAddress1 = "x_address1"
+	SRegAddress2 = "x_address2"
+	SRegCity     = "x_city"
+	SRegProvince = "x_province"
+	SRegPhone    = "x_phone"
+)
+
+const (
+	nsSReg  = "http://openid.net/extensions/sreg/1.1"
+	nsTeams = "http://ns.launchpad.net/2007/openid-teams"
+)
+
+var (
+	ErrCancel = errgo.New("login cancelled")
+	ErrOpenID = errgo.New("OpenID error")
```

Comment:
// ErrOpenID is the cause of an error returned from Verify
// that has been returned by the open-id server.

But... shouldn't this really be an error type that contains the
actual string returned? I guess it depends whether that
error string is intended to be something that programs can
inspect.